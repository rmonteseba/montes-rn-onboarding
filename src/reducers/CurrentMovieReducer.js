import { createSlice } from '@reduxjs/toolkit';

import { getUpcomingMovies, getPopularMovies, getTopRatedMovies } from '@/actions/MovieActions';
import { getCurrentMovie } from '@/actions/CurrentMovieActions';

const initialState = {
  currentMovie: {
    fulfilled: null,
    loading: false,
    error: null,
  },
};

const currentMovieSlice = createSlice({
  name: 'currentMovie',
  initialState,
  reducers: {
    clearCurrentMovie(state) {
      return {
        ...state,
        currentMovie: initialState.currentMovie,
      };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getCurrentMovie.pending, (state) => {
      state.currentMovie.loading = true;
      state.currentMovie.error = null;
      state.currentMovie.fulfilled = null;
    });
    builder.addCase(getCurrentMovie.rejected, (state, { payload }) => {
      state.currentMovie.loading = false;
      state.currentMovie.error = payload;
      state.currentMovie.fulfilled = null;
    });
    builder.addCase(getCurrentMovie.fulfilled, (state, { payload }) => {
      state.currentMovie.loading = false;
      state.currentMovie.error = null;
      state.currentMovie.fulfilled = payload;
    });
  },
});

const {
  reducer,
  actions: { clearCurrentMovie },
} = currentMovieSlice;

export { clearCurrentMovie, initialState };

export default reducer;
