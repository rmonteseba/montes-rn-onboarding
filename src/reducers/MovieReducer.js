import { createSlice } from '@reduxjs/toolkit';

import { getUpcomingMovies, getPopularMovies, getTopRatedMovies } from '@/actions/MovieActions';

const initialState = {
  topRatedMovies: {
    fulfilled: null,
    loading: false,
    error: null,
  },
  popularMovies: {
    fulfilled: null,
    loading: false,
    error: null,
  },
  upcomingMovies: {
    fulfilled: null,
    loading: false,
    error: null,
  },
};

const movieSlice = createSlice({
  name: 'movie',
  initialState,
  reducers: {
    clearTopRatedMovies(state) {
      return {
        ...state,
        topRatedMovies: initialState.topRatedMovies,
      };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getTopRatedMovies.pending, (state) => {
      state.topRatedMovies.loading = true;
      state.topRatedMovies.error = null;
      state.topRatedMovies.fulfilled = null;
    });
    builder.addCase(getTopRatedMovies.rejected, (state, { payload }) => {
      state.topRatedMovies.loading = false;
      state.topRatedMovies.error = payload;
      state.topRatedMovies.fulfilled = null;
    });
    builder.addCase(getTopRatedMovies.fulfilled, (state, { payload }) => {
      state.topRatedMovies.loading = false;
      state.topRatedMovies.error = null;
      state.topRatedMovies.fulfilled = payload; // TODO pages append
    });
    builder.addCase(getPopularMovies.pending, (state) => {
      state.popularMovies.loading = true;
      state.popularMovies.error = null;
      state.popularMovies.fulfilled = null;
    });
    builder.addCase(getPopularMovies.rejected, (state, { payload }) => {
      state.popularMovies.loading = false;
      state.popularMovies.error = payload;
      state.popularMovies.fulfilled = null;
    });
    builder.addCase(getPopularMovies.fulfilled, (state, { payload }) => {
      state.popularMovies.loading = false;
      state.popularMovies.error = null;
      state.popularMovies.fulfilled = payload; // TODO pages append
    });
    builder.addCase(getUpcomingMovies.pending, (state) => {
      state.upcomingMovies.loading = true;
      state.upcomingMovies.error = null;
      state.upcomingMovies.fulfilled = null;
    });
    builder.addCase(getUpcomingMovies.rejected, (state, { payload }) => {
      state.upcomingMovies.loading = false;
      state.upcomingMovies.error = payload;
      state.upcomingMovies.fulfilled = null;
    });
    builder.addCase(getUpcomingMovies.fulfilled, (state, { payload }) => {
      state.upcomingMovies.loading = false;
      state.upcomingMovies.error = null;
      state.upcomingMovies.fulfilled = payload; // TODO pages append
    });
  },
});

const {
  reducer,
  actions: { clearTopRatedMovies },
} = movieSlice;

export { clearTopRatedMovies };

export default reducer;
