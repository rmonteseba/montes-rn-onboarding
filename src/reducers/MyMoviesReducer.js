import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  myMovies: [],
};

const myMoviesSlice = createSlice({
  name: 'myMovies',
  initialState,
  reducers: {
    addMovie(state, { payload }) {
      return {
        ...state,
        myMovies: [...state.myMovies, payload],
      };
    },
    flushMovies(state) {
      return { ...state, myMovies: initialState.myMovies };
    },
  },
});

const {
  reducer,
  actions: { addMovie, flushMovies },
} = myMoviesSlice;

export { addMovie, flushMovies };

export default reducer;
