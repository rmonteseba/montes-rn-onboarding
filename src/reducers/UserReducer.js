import { createSlice } from '@reduxjs/toolkit';
import { login } from '@/actions/UserActions';

const initialState = {
  user: null,
  loading: false,
  error: null,
};

const userSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    logout(state) {
      return {
        ...state,
        error: initialState.error,
        loading: initialState.loading,
        user: initialState.user,
      };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(login.pending, (state) => {
      state.loading = true;
      state.error = null;
      state.user = null;
    });
    builder.addCase(login.rejected, (state, { payload }) => {
      state.loading = false;
      state.error = payload;
      state.user = null;
    });
    builder.addCase(login.fulfilled, (state, { payload }) => {
      state.loading = false;
      state.error = null;
      state.user = payload;
    });
  },
});

const {
  reducer,
  actions: { logout },
} = userSlice;

export { logout };

export default reducer;
