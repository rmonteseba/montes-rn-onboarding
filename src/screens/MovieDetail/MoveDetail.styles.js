import { StyleSheet } from 'react-native';
import { SCREEN_WIDTH } from '@/lib/utils/dimension-utils/dimensions';

export const baseStyles = StyleSheet.create({
  movieTitleStyle: {
    textAlign: 'center',
    fontSize: 35,
    marginVertical: 10,
  },
  movieOverviewStyle: {
    marginVertical: 10,
    textAlign: 'center',
    fontSize: 18,
  },
  movieGenreViewStyle: {
    flexDirection: 'row',
    marginVertical: 10,
    textAlign: 'center',
    alignSelf: 'center',
  },
  movieGenreStyle: {
    marginHorizontal: 10,
    textAlign: 'center',
    fontSize: 15,
  },
  imageStyles: {
    width: SCREEN_WIDTH,
    height: 400,
    backgroundColor: 'black',
  },
  spinnerTextStyle: {
    color: '#FFF',
  },
});
