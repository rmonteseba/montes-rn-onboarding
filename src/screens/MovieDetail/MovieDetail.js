import React, { useEffect } from 'react';
import { View, Text, Alert } from 'react-native';

import { ScrollView } from 'react-native-gesture-handler';
import { useRoute } from '@react-navigation/native';
import { useDispatch, useSelector } from 'react-redux';
import Spinner from 'react-native-loading-spinner-overlay';
import { baseStyles } from '@/screens/MovieDetail/MoveDetail.styles';
import { getCurrentMovie as getCurrentMovieSelector } from '@/selectors/CurrentMovieSelectors';
import { getCurrentMovie } from '@/actions/CurrentMovieActions';
import { strings } from '@/localization';
import MovieMainThumbnail from '@/components/main-movie/MovieMainThumbnail';

const MovieDetail = () => {
  const {
    params: { movieId },
  } = useRoute();

  const dispatch = useDispatch();

  const {
    error: currentMovieError,
    fulfilled: currentMovie,
    loading: currentMovieLoading,
  } = useSelector(getCurrentMovieSelector);

  useEffect(() => {
    if (movieId) {
      dispatch(getCurrentMovie({ movieId }));
    }
  }, [dispatch, movieId]);

  useEffect(() => {
    if (currentMovieError) {
      Alert.alert(currentMovieError);
    }
  }, [currentMovieError]);

  return (
    <ScrollView>
      <Spinner
        visible={currentMovieLoading}
        textContent={'Loading...'}
        textStyle={baseStyles.spinnerTextStyle}
      />
      {currentMovie && (
        <View style={baseStyles.container}>
          <Text style={baseStyles.movieTitleStyle}>{currentMovie.title}</Text>
          <MovieMainThumbnail
            imageStyles={baseStyles.imageStyles}
            path={currentMovie.poster_path}
          />
          <View style={baseStyles.movieGenreViewStyle}>
            {currentMovie?.genres?.map(({ id }) => (
              <Text style={baseStyles.movieGenreStyle} key={id}>
                {`\u29BF ${strings.movieDetail.genres[id]}`}
              </Text>
            ))}
          </View>
          <Text style={baseStyles.movieOverviewStyle}>{currentMovie.overview}</Text>
        </View>
      )}
    </ScrollView>
  );
};

export default MovieDetail;
