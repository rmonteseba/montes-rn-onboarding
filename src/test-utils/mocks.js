export const movie = {
  adult: false,
  backdrop_path: '/fBshLiEJcjdfrU3qQBIINcePSsm.jpg',
  belongs_to_collection: null,
  budget: 0,
  genres: [
    { id: 16, name: 'Animation' },
    { id: 878, name: 'Science Fiction' },
    { id: 12, name: 'Adventure' },
    { id: 10751, name: 'Family' },
  ],
  homepage: '',
  id: 877269,
  imdb_id: 'tt10298840',
  original_language: 'en',
  original_title: 'Strange World',
  overview:
    'A journey deep into an uncharted and treacherous land, where fantastical creatures await the legendary Clades—a family of explorers whose differences threaten to topple their latest, and by far most crucial, mission.',
  popularity: 63.426,
  poster_path: '/a9x70QUXs6uEwtin1SGNKoFXoia.jpg',
  production_companies: [
    {
      id: 2,
      logo_path: '/wdrCwmRnLFJhEoH8GSfymY85KHT.png',
      name: 'Walt Disney Pictures',
      origin_country: 'US',
    },
    {
      id: 6125,
      logo_path: '/tzsMJBJZINu7GHzrpYzpReWhh66.png',
      name: 'Walt Disney Animation Studios',
      origin_country: 'US',
    },
  ],
  production_countries: [{ iso_3166_1: 'US', name: 'United States of America' }],
  release_date: '2022-11-23',
  revenue: 0,
  runtime: 102,
  spoken_languages: [
    { english_name: 'English', iso_639_1: 'en', name: 'English' },
    { english_name: 'Slovak', iso_639_1: 'sk', name: 'Slovenčina' },
  ],
  status: 'In Production',
  tagline: 'Journey to a place where nothing is as it appears!',
  title: 'Strange World',
  video: false,
  vote_average: 0.0,
  vote_count: 0,
};

export const movieResponse = movie;

export const movies = [
  {
    adult: false,
    backdrop_path: '/rSPw7tgCH9c6NqICZef4kZjFOQ5.jpg',
    genre_ids: [18, 80],
    id: 238,
    original_language: 'en',
    original_title: 'The Godfather',
    overview:
      'Spanning the years 1945 to 1955, a chronicle of the fictional Italian-American Corleone crime family. When organized crime family patriarch, Vito Corleone barely survives an attempt on his life, his youngest son, Michael steps in to take care of the would-be killers, launching a campaign of bloody revenge.',
    popularity: 105.925,
    poster_path: '/3bhkrj58Vtu7enYsRolD1fZdja1.jpg',
    release_date: '1972-03-14',
    title: 'The Godfather',
    video: false,
    vote_average: 8.7,
    vote_count: 16846,
  },
  {
    adult: false,
    backdrop_path: '/kXfqcdQKsToO0OUXHcrrNCHDBzO.jpg',
    genre_ids: [18, 80],
    id: 278,
    original_language: 'en',
    original_title: 'The Shawshank Redemption',
    overview:
      'Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.',
    popularity: 88.027,
    poster_path: '/q6y0Go1tsGEsmtFryDOJo3dEmqu.jpg',
    release_date: '1994-09-23',
    title: 'The Shawshank Redemption',
    video: false,
    vote_average: 8.7,
    vote_count: 22585,
  },
  {
    adult: false,
    backdrop_path: '/aVFx1VtlOxR3v0ADEatalXOvwbu.jpg',
    genre_ids: [16, 14, 28],
    id: 620249,
    original_language: 'zh',
    original_title: '罗小黑战记',
    overview:
      'In the bustling human world, spirits live peacefully with mankind. Luo Xiaohei, a young cat spirit, becomes a nomad after his forest home is destroyed. His natural talent soon attracts spirits and humans alike, but all of them have their own motivations.',
    popularity: 19.13,
    poster_path: '/aLv87NgRJUPkQ6sVLP72IisDdt4.jpg',
    release_date: '2019-08-27',
    title: 'The Legend of Hei',
    video: false,
    vote_average: 8.7,
    vote_count: 200,
  },
];

export const moviesResponse = {
  page: 1,
  results: movies,
  total_pages: 524,
  total_results: 10469,
};
