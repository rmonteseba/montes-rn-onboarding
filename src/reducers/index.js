import { combineReducers } from 'redux';
import { persistReducer } from 'redux-persist';
import { storage } from '@/storage';
import { errorReducer } from '@/reducers/ErrorReducer';
import { statusReducer } from '@/reducers/StatusReducer';
import userReducer from '@/reducers/UserReducer';
import movieReducer from '@/reducers/MovieReducer';
import myMoviesReducer from '@/reducers/MyMoviesReducer';
import currentMovieReducer from '@/reducers/CurrentMovieReducer';

const userPersistConfig = {
  key: 'user',
  storage,
  whitelist: ['user'],
};

const moviePersistConfig = {
  key: 'movie',
  storage,
};

const myMoviesPersistConfig = {
  key: 'myMovies',
  storage,
};

export const rootReducer = combineReducers({
  error: errorReducer,
  status: statusReducer,
  user: persistReducer(userPersistConfig, userReducer),
  movie: persistReducer(moviePersistConfig, movieReducer),
  myMovies: persistReducer(myMoviesPersistConfig, myMoviesReducer),
  currentMovie: currentMovieReducer,
});
