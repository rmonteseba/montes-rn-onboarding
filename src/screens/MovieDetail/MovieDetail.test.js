import { render, screen, waitFor } from '@testing-library/react-native';
import React from 'react';
import MockAdapter from 'axios-mock-adapter';
import axios from '@/networking/clients/movieDBClient';
import MovieDetail from '@/screens/MovieDetail/MovieDetail';
import { withProviders } from '@/test-utils';
import { movieResponse } from '@/test-utils/mocks';
import { initialState as currentMovieInitialState } from '@/reducers/CurrentMovieReducer';

const mockedParams = { movieId: movieResponse.id };

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useRoute: () => ({
      params: mockedParams,
    }),
  };
});

const fakeStore = {
  error: {},
  status: {},
  user: {
    username: 'johndoe',
  },
  currentMovie: currentMovieInitialState,
};

describe('MovieDetail', () => {
  let axiosMock;

  beforeEach(() => {
    axiosMock = new MockAdapter(axios);
    axiosMock.onGet(`/movie/${movieResponse.id}`).reply(200, movieResponse);
  });

  afterEach(() => {
    axiosMock.restore();
  });

  it('should match the snapshot', () => {
    const { toJSON } = render(withProviders(<MovieDetail />, { initialState: fakeStore }));
    expect(toJSON()).toMatchSnapshot();
  });

  it('should show the movie when the movie is loaded', async () => {
    const { findByText, toJSON } = render(
      withProviders(<MovieDetail />, {
        initialState: fakeStore,
      })
    );

    const movieTitle = await findByText(movieResponse.title);

    expect(movieTitle).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
});
