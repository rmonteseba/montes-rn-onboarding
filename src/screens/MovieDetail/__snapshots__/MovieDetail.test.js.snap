// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetail should match the snapshot 1`] = `
<RCTScrollView
  collapsable={false}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  waitFor={
    Array [
      Object {
        "current": null,
      },
    ]
  }
>
  <View>
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      statusBarTranslucent={true}
      supportedOrientations={
        Array [
          "landscape",
          "portrait",
        ]
      }
      transparent={true}
      visible={true}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "flex": 1,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.25)",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <ActivityIndicator
            color="white"
            size="large"
            style={
              Array [
                Object {
                  "flex": 1,
                },
                Object {},
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "flex": 1,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "height": 50,
                    "top": 80,
                  },
                  Object {
                    "color": "#FFF",
                  },
                ]
              }
            >
              Loading...
            </Text>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</RCTScrollView>
`;

exports[`MovieDetail should show the movie when the movie is loaded 1`] = `
<RCTScrollView
  collapsable={false}
  handlerTag={2}
  handlerType="NativeViewGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  waitFor={
    Array [
      Object {
        "current": null,
      },
    ]
  }
>
  <View>
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      statusBarTranslucent={true}
      supportedOrientations={
        Array [
          "landscape",
          "portrait",
        ]
      }
      transparent={true}
      visible={false}
    />
    <View>
      <Text
        style={
          Object {
            "fontSize": 35,
            "marginVertical": 10,
            "textAlign": "center",
          }
        }
      >
        Strange World
      </Text>
      <View
        accessibilityIgnoresInvertColors={true}
        style={
          Object {
            "backgroundColor": "black",
            "height": 400,
            "width": 750,
          }
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          resizeMode="contain"
          source={
            Object {
              "uri": "undefinedoriginal/a9x70QUXs6uEwtin1SGNKoFXoia.jpg",
            }
          }
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": 400,
                "width": 750,
              },
              undefined,
            ]
          }
        />
        <BVLinearGradient
          colors={
            Array [
              undefined,
              4294111986,
            ]
          }
          endPoint={
            Object {
              "x": 0,
              "y": 1,
            }
          }
          locations={null}
          startPoint={
            Object {
              "x": 0,
              "y": 0.4,
            }
          }
          style={
            Object {
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "flexDirection": "row",
            "marginVertical": 10,
            "textAlign": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 15,
              "marginHorizontal": 10,
              "textAlign": "center",
            }
          }
        >
          ⦿ Animation
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "marginHorizontal": 10,
              "textAlign": "center",
            }
          }
        >
          ⦿ Science Fiction
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "marginHorizontal": 10,
              "textAlign": "center",
            }
          }
        >
          ⦿ Adventure
        </Text>
        <Text
          style={
            Object {
              "fontSize": 15,
              "marginHorizontal": 10,
              "textAlign": "center",
            }
          }
        >
          ⦿ Family
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 18,
            "marginVertical": 10,
            "textAlign": "center",
          }
        }
      >
        A journey deep into an uncharted and treacherous land, where fantastical creatures await the legendary Clades—a family of explorers whose differences threaten to topple their latest, and by far most crucial, mission.
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
